File: PythonSandbox/src/leetcode/lc141_linked_list_cycle.py
class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None

class Solution:
    def hasCycle(self, head) -> bool:
        if head == None:
            return False

        idSet = set()

        n = head
        while n != None:
            
            nid = id(n)

            if nid in idSet:
                return True
            else:
                idSet.add(nid)

            n = n.next
        
        return False

File: JavaSandbox/src/leetcode/LC141_LinkedListCycle.java
package leetcode;

public class LC141_LinkedListCycle {
    public boolean hasCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while(slow != null && fast != null) {
            
            fast = fast.next;

            if(fast == null) {
                return false;
            }

            fast = fast.next;
            slow = slow.next;

            if(slow == fast) {
                return true;
            }
        }

        return false;
    }
}

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}